### ISSUE TITLE
Inner Join or EF Include equivalente

### ISSUE THREAD

**Issue by EnderXXX** [RobThree/MongoRepository#15]
Hi, I'm wondering how to be able to join two or more Repositories result by common properties (normally foreign pks) using this wonderful library. As fas as I know it could be done with EF and SQL but I am very happy with NoSQL. 

As MongoRepository seems to implement IQueryable, do you think that something like the next code could work? (if in a fluent way were possible even better^^) thanks!

"from w in repo.Query<Worker>()
join e in repo.Query<XEntity>() on.."

**Comment by RobThree** [comment]
Just as with other (you could say, "more classical") databases you need to optimize the way data is stored to your needs but also in such a way that it is not an unmaintainable mess of data everywhere. So it really depends on your situation on what the best (or better: best balanced) solution would be.

You can imagine a collection of `customer`s and a collection of `product`s and a collection of `shopping cart`s that contains a link between a `customer` and one or more `product`s. This is fairly close to what you'd do in an RDBMS and, indeed, use "joins" for to gather the information.

In Mongo's world, however, you would more likely have a collection of `product`s and `customer`s and **as part of the customer** you'd store a `shoppingcart` (or: N-products in a `cart` group) as part of the `customer` object itself. This would only have 'references' to the `product`s in the respective collection. That would mean "one less join".

Mongo does have concepts similar to joins (and the, [to be released, 3.2 version](https://www.mongodb.com/mongodb-3.2) promises even more "alike" concepts) but the general idea is that an "extra roundtrip to the server" isn't that bad in document oriented systems like Mongo as it is for "classic RDBMS'es" (and please don't quote me on this, I'm trying very hard not to over/understate things or make it too black-and-white; it's actually in all [database models](https://en.wikipedia.org/wiki/Database_model) very much more of a grey area than it may sound in what I'm saying).

So you can, already, do joins as Linq provides on IQueryables, however, they *may* not do exactly what you *think* they do (but that al very much depends on the implementation in, amongst others, the underlying driver).

I have tried to demonstrate below:

```c#
using MongoRepository;
using System.Collections.Generic;
using System.Linq;

class Customer : Entity
{
    public string Name { get; set; }
    public List<string> Cart { get; set; }
}

class Product : Entity
{
    public string Description { get; set; }
    public decimal Price { get; set; }
}


class Program
{
    static void Main(string[] args)
    {
        var custrepo = new MongoRepository<Customer>();
        var prodrepo = new MongoRepository<Product>();


        // Set up some entities:
        var soap = new Product { Description = "Hand soap", Price = 1.23m };
        var milk = new Product { Description = "Milk", Price = 1.99m };
        var chips = new Product { Description = "Chips", Price = .99m };
        var bread = new Product { Description = "Bread", Price = 2.10m };
        prodrepo.Add(new[] { soap, milk, chips, bread });

        // Now a customer comes along...
        var rob = new Customer { Name = "Rob Janssen" };

        // ...and adds some products into two carts
        rob.Cart = new List<string>() { soap.Id, bread.Id };
        custrepo.Add(rob);

        // Essentially, now, when we retrieve a customer we also get the contents of the cart; no join required
        var cust = custrepo.Where(c => c.Name == "Rob Janssen").First();


        // Let's retrieve all products in the customer's cart:
        var products = from cartitem in cust.Cart
                       join product in prodrepo on cartitem equals product.Id
                       select product;
    }
}
```

This demonstrates a 1:N stored **as part of the customer** (namely the cart's contents) but also a 1:N stored separately (requiring "a join", namely the products).

You *could* go as far as storing the products (or a copy of a product) in the cart:

```c#
using MongoRepository;
using System.Collections.Generic;
using System.Linq;

class Customer : Entity
{
    public string Name { get; set; }
    public List<Product> Cart { get; set; }
}

class Product : Entity
{
    public string Description { get; set; }
    public decimal Price { get; set; }
}


class Program
{
    static void Main(string[] args)
    {
        var custrepo = new MongoRepository<Customer>();
        var prodrepo = new MongoRepository<Product>();

        custrepo.DeleteAll();
        prodrepo.DeleteAll();

        // Set up some entities:
        var soap = new Product { Description = "Hand soap", Price = 1.23m };
        var milk = new Product { Description = "Milk", Price = 1.99m };
        var chips = new Product { Description = "Chips", Price = .99m };
        var bread = new Product { Description = "Bread", Price = 2.10m };
        prodrepo.Add(new[] { soap, milk, chips, bread });

        // Now a customer comes along...
        var rob = new Customer { Name = "Rob Janssen" };

        // ...and adds some products into two carts
        rob.Cart = new List<Product>() { soap, bread };
        custrepo.Add(rob);

        // Essentially, now, when we retrieve a customer we also get the contents of the cart; no join required
        var cust = custrepo.Where(c => c.Name == "Rob Janssen").First();
    }
}
```

However, you can see that this results in some 'duplication' you probably don't want. Then again; maybe you do. This totally depends on the situation, use-case etc.

**Comment by RobThree** [comment]
You're welcome!
